pcdsdaq/daq/lcls2.py

```diff
@@ -0,0 +1,1780 @@
+  and through specific transitions. This means we can schedule
+  code to be run when the daq reaches specific states e.g. when
+  it hits running or when it undergoes specific transitions e.g.
+  endrun, or more simply we can wait for specific states and
+  transitions.
+- DaqLCLS2.state_transition is the main communication point for
+  sending state to the DAQ. This is where we pass all information
+  to the DAQ with the exception of the record/no record
+  configuration. We assemble information like the extra controls
+  and the events for the step length here and make sure the DAQ
+  recieves them, as well as most of the cosmetic DAQ
+  configuration values. Ultimately this method causes the DAQ
+  to transition, e.g. to configure itself, to start a run,
+  to pause a run, etc.
+- DaqLCLS2.configure is the only place where we might find
+  ourselves setting the DAQ's recording state (if requested).
+
+See the following URL for a description of the DAQ state machine:
+https://confluence.slac.stanford.edu/display/PSDMInternal/Finite+State+Machine
+"""
+from __future__ import annotations
+
+import logging
+import threading
+import time
+from functools import cache
+from numbers import Real
+from typing import Any, Iterator, Optional, Union, get_type_hints
+
+from bluesky import RunEngine
+from ophyd.device import Component as Cpt
+from ophyd.signal import Signal
+from ophyd.status import Status
+from ophyd.utils import StatusTimeoutError, WaitTimeoutError
+from ophyd.utils.errors import InvalidState
+from psdaq.control.ControlDef import ControlDef
+from psdaq.control.DaqControl import DaqControl
+
+from .interface import (CONFIG_VAL, ControlsArg, DaqBase,
+                        DaqStateTransitionError, DaqTimeoutError, EnumId,
+                        HelpfulIntEnum, Sentinel, TernaryBool,
+                        get_controls_name, get_controls_value, typing_check)
+
+logger = logging.getLogger(__name__)
+
+
+class DaqLCLS2(DaqBase):
+    """
+    The LCLS2 DAQ as a bluesky-compatible object.
+
+    This uses the ``psdaq.control.DaqControl`` module to send ZMQ commands
+    to the DAQ instance.
+
+    It can be used as a ``Reader`` or as a ``Flyer`` in a ``bluesky`` plan.
+
+    Parameters
+    ----------
+    platform : int
+        Required argument to specify which daq platform we're connecting to.
+    host : str
+        The hostname of the DAQ host we are connecting to.
+    timeout : int
+        How many milliseconds to wait for various DAQ communications before
+        reporting an error.
+    RE: ``RunEngine``, optional
+        Set ``RE`` to the session's main ``RunEngine``
+    hutch_name: str, optional
+        Define a hutch name to use instead of shelling out to get_hutch_name.
+    sim: bool, optional
+        If True, simulate the daq. Defaults to False.
+    """
+    step_value_sig = Cpt(Signal, value=1, kind='normal')
+    state_sig = Cpt(Signal, value=None, kind='normal')
```

Comment:
Doc kwargs would be nice, though not truly necessary:
```suggestion
    state_sig = Cpt(Signal, value=None, kind='normal', doc="Current state (state_enum)")
```

Reply:
I forgot this was a thing, I'll go through and add them.